### PR TITLE
Added 2nd training workflow: data generator on root files

### PR DESCRIPTION
**Pull request by hanhiller**
Added 'DataGenerator.py', which generates batches of data from root files in the specified (input) directory. Then in 'train.py', three instances (training, valid, and testing) of this class are called to create the feed data, batch by batch, into the keras model and perform the subsequent analysis.

Additionally, when executing 'python train.py', you now specify whether you want train from the root files directly (using the data generator) or convert the root files to h5 and train from these (reading all the data at once). The h5 conversion is now executed within 'train.py.

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -11,8 +11,9 @@ dependencies:
   - pandas
   - h5py
   - tqdm
+  - keras
```
Comment (jmduarte) [inline environment.yml:14]: remove this

File: train.py [hunk 2]
```diff
@@ -50,8 +209,26 @@ def main(args):
         print ('Creating directory' + path_out)
 
     # Read inputs
+    
+    # convert root files to h5 and store in same location
+    i =0
+    for file in os.listdir(inputPath):
```
Comment (jmduarte) [inline train.py:215]: could do something like
```python
for i, file in enumerate(glob.glob(os.path.join(inputPath,'*.root'))):
```

File: train.py [hunk 3]
```diff
@@ -26,10 +28,166 @@
...
+
+    fi.write("Working Time (s) : {}".format(end_time - start_time))
+    fi.write("Working Time (m) : {}".format((end_time - start_time)/60.))
+
+    fi.close()
+
+
+# Configuration
+
+'''if __name__ == "__main__":
```
Comment (jmduarte) [inline train.py:175]: remove commented code

File: train.py [hunk 4]
```diff
@@ -76,12 +253,12 @@ def main(args):
...
-    print("# of events higher than 300 GeV : {}".format(Yr_pt.shape[0]))
+    #print("# of events higher than 300 GeV : {}".format(Yr_pt.shape[0]))
 
     indices = np.array([i for i in range(len(Yr))])
-    print(indices)
-    indices_train, indices_test = train_test_split(indices, test_size=0.2, random_state= 7)
-    indices_train, indices_valid = train_test_split(indices_train, test_size=0.2, random_state=7)
+    #print(indices)
+    indices_train, indices_valid = train_test_split(indices, test_size=0.2, random_state= 7)
+    indices_train, indices_test = train_test_split(indices_train, test_size=0.2, random_state=7)
```
Comment (jmduarte) [inline train.py:261]: i'm confused... should it be reversed?

usually, we split the data into training+validation and testing (separate)

in the end it probably doesn't matter much

File: train.py [hunk 5]
```diff
@@ -189,7 +362,7 @@ def main(args):
 
 # Configuration
 
-if __name__ == "__main__":
+''' if __name__ == "__main__":
```
Comment (jmduarte) [inline train.py:365]: remove commented code

File: utils.py [hunk 6]
```diff
@@ -150,3 +150,6 @@ def Make1DHists(truth, predict, PUPPI, xmin=0, xmax=400, nbins=100, density=Fals
     plt.ylabel(yname)
     plt.savefig(outputname)
     plt.close()
+    
+def to_np_array(ak_array, maxN=100, pad=0):
```
Comment (jmduarte) [inline utils.py:154]: reuse this function in `convert.py`

**Review by jmduarte — CHANGES_REQUESTED** [review]

File: train.py [hunk 7]
```diff
@@ -50,8 +209,26 @@ def main(args):
         print ('Creating directory' + path_out)
 
     # Read inputs
+    
+    # convert root files to h5 and store in same location
+    i =0
+    for file in os.listdir(inputPath):
+        if '.root' in file:
+            os.system(f'python convertNanoToHDF5_L1triggerToDeepMET.py -i {inputPath}/{file} -o {inputPath}/set{i}.h5')
```
Comment (jmduarte) [inline train.py:217]: could maybe chack if the h5 files already exist before converting them

File: utils.py (hunk 6 above)
Comment (hanhiller) [reply utils.py:154]: You're saying I should just define in `convert.py`, right?

Comment (jmduarte) [reply utils.py]: No, you can keep the definition here but just import it and use it in `convert.py`

**Review by jmduarte — APPROVED** [review]

**Pull request merged**